athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/GcsUtil.java

```diff
@@ -0,0 +1,192 @@
+    {
+        com.amazonaws.services.glue.model.GetTableRequest getTableRequest = new com.amazonaws.services.glue.model.GetTableRequest();
+        getTableRequest.setDatabaseName(tableName.getSchemaName());
+        getTableRequest.setName(tableName.getTableName());
+
+        GetTableResult result = awsGlue.getTable(getTableRequest);
+        return result.getTable();
+    }
+
+    public static Object coerce(FieldVector vector, Object value)
+    {
+        switch (vector.getMinorType()) {
+            case TIMESTAMPNANO:
+            case TIMENANO:
+                if (value instanceof org.joda.time.LocalDateTime) {
+                    DateTimeZone dtz = ((org.joda.time.LocalDateTime) value).getChronology().getZone();
+                    return ((org.joda.time.LocalDateTime) value).toDateTime(dtz).getMillis();
+                }
+                if (value instanceof ZonedDateTime) {
+                    return DateTimeFormatterUtil.packDateTimeWithZone((ZonedDateTime) value);
+                }
+                else if (value instanceof LocalDateTime) {
+                    return DateTimeFormatterUtil.packDateTimeWithZone(
+                            ((LocalDateTime) value).atZone(UTC_ZONE_ID).toInstant().toEpochMilli(), UTC_ZONE_ID.getId());
+                }
+                else if (value instanceof Date) {
+                    long ldtInLong = Instant.ofEpochMilli(((Date) value).getTime())
+                            .atZone(UTC_ZONE_ID).toInstant().toEpochMilli();
+                    return DateTimeFormatterUtil.packDateTimeWithZone(ldtInLong, UTC_ZONE_ID.getId());
+                }
+                else {
+                    return (long) value / 1000000;
+                }
+            case TIMEMICRO:
+            case TIMESTAMPMICRO:
+            case TIMESTAMPMICROTZ:
+                if (value instanceof org.joda.time.LocalDateTime) {
+                    DateTimeZone dtz = ((org.joda.time.LocalDateTime) value).getChronology().getZone();
+                    return ((org.joda.time.LocalDateTime) value).toDateTime(dtz).getMillis();
+                }
+                if (value instanceof ZonedDateTime) {
+                    return DateTimeFormatterUtil.packDateTimeWithZone((ZonedDateTime) value);
+                }
+                else if (value instanceof LocalDateTime) {
+                    return DateTimeFormatterUtil.packDateTimeWithZone(
+                            ((LocalDateTime) value).atZone(UTC_ZONE_ID).toInstant().toEpochMilli(), UTC_ZONE_ID.getId());
+                }
+                else if (value instanceof Date) {
+                    long ldtInLong = Instant.ofEpochMilli(((Date) value).getTime())
+                            .atZone(UTC_ZONE_ID).toInstant().toEpochMilli();
+                    return DateTimeFormatterUtil.packDateTimeWithZone(ldtInLong, UTC_ZONE_ID.getId());
+                }
+                else {
+                    return (long) value / 1000;
+                }
```

Comment:
These should not all be lumped together under the `TIMESTAMPMICROTZ` case.

`TIMESTAMPMICROTZ` has a timezone, so it should only be dealing with the situations where there is a timezone (ZonedDateTime and Date).

`TIMEMICRO` and `TIMESTAMPMICRO` don't have timezones, so those should be dealing with the situations where there isn't a timezone (like LocalDateTime and just a raw long).

And in the raw long case for microseconds, do this instead:
```
TimeUnit.MICROSECONDS.convert(value, TimeUnit.MILLISECONDS);
// Duration doesn't support microseconds
```

Reply:
Additionally, why are we expecting joda time types at all?

Do we actually ever see the `Object value` being a joda time type other than maybe in your tests?

I grepped through arrow's java code and I don't see anything referring to joda time other than in the Flight SQL library which we aren't using.

Reply:
> This is a holiday for me so I'm done commenting for today. Some general comments:
> 
> * How are joda time, LocalDateTime, DateTime, etc kind of types showing up as Arrow values? Arrow Dataset is giving you Arrow vectors right? As far as I understand, the Arrow types for dates and time, are just longs + some Arrow type, right?

we are getting actual values from Dataset API in the record handler in the case of date and time, not only longs.
we were checking joda time types just to be double sure, we had not got any jodatype while testing.

Reply:
My point is that if you are not getting these other types, we shouldn't be writing code to convert those types that we'll never get.
Are you saying that we don't get joda types but we do get DateTime types for the field values?

Reply:
yes. 
![image](https://user-images.githubusercontent.com/108925344/213195380-8af82089-38f9-4640-8ffa-62569046d31b.png)

Reply:
Ok, thanks for confirming this. I will have to do a deeper dive to figure out what Arrow is doing.